OpenSSL/SSL.py

```diff
@@ -1030,6 +1030,45 @@ def recv(self, bufsiz, flags=None):
     read = recv
 
 
+    def recv_into(self, buffer, nbytes=None, flags=None):
+        """
+        Receive data on the connection and store the data into a buffer rather
+        than creating a new string.
+
+        :param buffer: The buffer to copy into.
+        :param nbytes: (optional) The maximum number of bytes to read into the
+            buffer. If not present, defaults to the size of the buffer. If
+            larger than the size of the buffer, is reduced to the size of the
+            buffer.
+        :param flags: (optional) Included for compatibility with the socket
+            API, the value is ignored.
+        :return: The number of bytes read into the buffer.
+        """
+        if nbytes is None:
+            nbytes = len(buffer)
+        else:
+            nbytes = min(nbytes, len(buffer))
+
+        # We need to create a temporary buffer. This is annoying, it would be
+        # better if we could pass memoryviews straight into the SSL_read call,
+        # but right now we can't. Revisit this if CFFI gets that ability.
+        buf = _ffi.new("char[]", nbytes)
```

Comment:
Yes, I believe that's correct.

On Sun, Mar 15, 2015 at 6:05 PM, Jean-Paul Calderone <
notifications@github.com> wrote:

> In OpenSSL/SSL.py
> <https://github.com/pyca/pyopenssl/pull/199#discussion_r26453014>:
>
> > +            buffer. If not present, defaults to the size of the buffer. If
> > +            larger than the size of the buffer, is reduced to the size of the
> > +            buffer.
> > +        :param flags: (optional) Included for compatibility with the socket
> > +            API, the value is ignored.
> > +        :return: The number of bytes read into the buffer.
> > +        """
> > +        if nbytes is None:
> > +            nbytes = len(buffer)
> > +        else:
> > +            nbytes = min(nbytes, len(buffer))
> > +
> > +        # We need to create a temporary buffer. This is annoying, it would be
> > +        # better if we could pass memoryviews straight into the SSL_read call,
> > +        # but right now we can't. Revisit this if CFFI gets that ability.
> > +        buf = _ffi.new("char[]", nbytes)
>
> Looks like only in some cases. Not if the given buffer is a bytearray,
> for example?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pyca/pyopenssl/pull/199/files#r26453014>.
>



-- 
"I disapprove of what you say, but I will defend to the death your right to
say it." -- Evelyn Beatrice Hall (summarizing Voltaire)
"The people's good is the highest law." -- Cicero
GPG Key fingerprint: 125F 5C67 DFE9 4084